### PR TITLE
fix: imagestream updates doesn't trigger Deployment

### PR DESCRIPTION
**Pull request by gijohn**
Fixes: https://jira.coreos.com/browse/ODC-2128

### CODE REVIEW

File: frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts [hunk 1]
```diff
@@ -95,28 +95,51 @@ const getMetadata = (formData: DeployImageFormData) => {
...
+    isi: { image, ports },
     deployment: { env, replicas },
     labels: userLabels,
     limits: { cpu, memory },
+    imageStream: { image: imgName, namespace: imgNamespace, tag },
   } = formData;
 
+  const defaultAnnotations = {
+    ...annotations,
+    ...{
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts]: There really isn't any value to spreading an object of static values. Just place them in one by one after the spread.

ie
```
{
  ...someOtherObject,
  keyA: valueA,
  keyInOtherObject: newValue,
  etc: etc,
}
```

File: frontend/packages/dev-console/src/components/import/import-submit-utils.ts [hunk 2]
```diff
@@ -217,7 +216,18 @@ export const createDeployment = (
 
   const imageStreamName = imageStream && imageStream.metadata.name;
   const defaultLabels = getAppLabels(name, application, imageStreamName, tag);
-  const defaultAnnotations = getAppAnnotations(repository, ref);
+  const annotations = {
+    ...getAppAnnotations(repository, ref),
+    ...{
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/import-submit-utils.ts]: Same here, no need to spread, just define them directly after the `getAppAnnotations` spread

File: frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts [hunk 3]
```diff
@@ -95,28 +95,51 @@ const getMetadata = (formData: DeployImageFormData) => {
...
+
   const deployment = {
     kind: 'Deployment',
     apiVersion: 'apps/v1',
     metadata: {
       name,
       namespace,
       labels,
-      annotations,
+      defaultAnnotations,
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts]: Is this really supposed to be `defaultAnnotations`? 

I see the data going up but the echo back from the server doesn't have this property in the metadata... I assume you want to be adding actual annotations?

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts]: Ah - yeah the import utils work fine, this appears to be a bug.

File: frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts [hunk 4]
```diff
@@ -46,4 +51,108 @@ describe('DeployImage Submit Utils', () => {
...
+      spyOn(k8s, 'k8sCreate').and.callFake((model, data, dryRun) => ({ model, data, dryRun }));
+    });
+
+    afterAll(() => {
+      jest.restoreAllMocks();
+    });
+
+    it('should choose image from dockerImageReference when creating Deployment using internal imagestream', () => {
+      const returnValue = createDeployment(internalImageData, false);
+      expect(_.get(returnValue, 'model.kind')).toEqual('Deployment');
```
[the remark this answers is not in the document]
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts]: done

File: frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts [hunk 5]
```diff
@@ -46,4 +51,108 @@ describe('DeployImage Submit Utils', () => {
       expect(values.isi.ports).toContainEqual({ containerPort: 8081, protocol: 'TCP' });
     });
   });
+
+  describe('createDeployment tests', () => {
+    beforeAll(() => {
+      spyOn(k8s, 'k8sCreate').and.callFake((model, data, dryRun) => ({ model, data, dryRun }));
```
[the remark this answers is not in the document]
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts]: if I just use spyOn, it defaults to Jasmine and it was throwing type error. I have to specifically use jest.spyOn.

Done

File: frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts [hunk 6]
```diff
@@ -46,4 +51,108 @@ describe('DeployImage Submit Utils', () => {
...
+        limits: { cpu: '10m', memory: '200Mi' },
+        requests: { cpu: '5m', memory: '100Mi' },
+      });
+    });
+  });
+
+  describe('createResource tests', () => {
+    beforeAll(() => {
+      spyOn(k8s, 'k8sCreate').and.callFake((model, data, dryRun) =>
+        Promise.resolve({ model, data, dryRun }),
```
[the remark this answers is not in the document]
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts]: done

File: frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts [hunk 7]
```diff
@@ -46,4 +51,108 @@ describe('DeployImage Submit Utils', () => {
...
+        .catch(() => {
+          done();
+        });
+    });
+
+    it('should call KNative when creating Resources when resource is KNative', (done) => {
+      const mockData = _.cloneDeep(defaultData);
+      mockData.resources = Resources.KnativeService;
+
+      const imageStreamSpy = spyOn(submitUtils, 'createImageStream').and.callFake(() => ({
```
[the remark this answers is not in the document]
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts]: done

File: frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+import { Resources } from '../import-types';
+import * as submitUtils from '../import-submit-utils';
+import { defaultData, nodeJsBuilderImage as buildImage } from './import-submit-utils-data';
+
+const { createDeployment, createResources } = submitUtils;
+
+describe('Import Submit Utils', () => {
+  describe('createDeployment tests', () => {
+    beforeAll(() => {
+      spyOn(k8s, 'k8sCreate').and.callFake((model, data, dryRun) => ({ model, data, dryRun }));
```
[the remark this answers is not in the document]
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts]: done

File: frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        limits: { cpu: '10m', memory: '200Mi' },
+        requests: { cpu: '5m', memory: '100Mi' },
+      });
+    });
+  });
+
+  describe('createResource tests', () => {
+    beforeAll(() => {
+      spyOn(k8s, 'k8sCreate').and.callFake((model, data, dryRun) =>
+        Promise.resolve({ model, data, dryRun }),
```
[the remark this answers is not in the document]
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts]: done

File: frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts (hunk 3 above)
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts]: good catch. it should be `annotations: defaultAnnotations`. Changed.

File: frontend/packages/dev-console/src/components/import/import-submit-utils.ts (hunk 2 above)
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/import-submit-utils.ts]: done

File: frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts (hunk 1 above)
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts]: done

File: frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts [hunk 10]
```diff
@@ -46,4 +58,117 @@ describe('DeployImage Submit Utils', () => {
       expect(values.isi.ports).toContainEqual({ containerPort: 8081, protocol: 'TCP' });
     });
   });
+
+  describe('createDeployment tests', () => {
+    beforeAll(() => {
+      jest
+        .spyOn(k8s, 'k8sCreate')
+        .mockImplementation((model, data, dryRun) => ({ model, data, dryRun }));
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts]: Why some with `Promise` and some without?

File: frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+import { defaultData, nodeJsBuilderImage as buildImage } from './import-submit-utils-data';
+
+const { createDeployment, createResources } = submitUtils;
+
+describe('Import Submit Utils', () => {
+  describe('createDeployment tests', () => {
+    beforeAll(() => {
+      jest
+        .spyOn(k8s, 'k8sCreate')
+        .mockImplementation((model, data, dryRun) => ({ model, data, dryRun }));
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts]: Same here, shouldn't this be returning a wrapped `Promise`?

File: frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts (hunk 10 above)
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/deployImage-submit-utils.spec.ts]: changed

File: frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts (hunk 11 above)
Comment (gijohn) [reply frontend/packages/dev-console/src/components/import/__tests__/import-submit-utils.spec.ts]: changed

**Pull request merged**